pl-faded-parsons.js

```diff
@@ -1,377 +1,720 @@
+            // line moved to output and logged there
+            return;
+          }
+          widget.addLogEntry(
+            { type: "moveInput", target: ui.item[0].id },
+            true,
+          );
+        },
+      });
+      solutionTray.sortable("option", "connectWith", starterTray);
     }
-    markCorrect() {}
-    markIncorrectPosition() {}
-    markIncorrectIndent() {}
-}
 
-class ParsonsWidget {
-    constructor(options) {
-        const defaults = {
-            source_tray: false,
-            source_tray: "starter-code",
-            solution_tray: "parsons-solution",
-            x_indent: 4,
-            can_indent: true,
-            onSortableUpdate: (_event, _ui) => {},
-            onBlankUpdate: (_event, _codeline) => {},
-
-            // TODO: reimplement support (tagged with TODO:FB)
-            incorrectSound: false,
-            feedback_cb: false,
-            first_error_only: true,
-            max_wrong_lines: 10, // maybe not needed?
-            lang: "en",
+    widget.redrawTabStops();
+
+    $(widget.config.main)
+      .find("li.codeline")
+      .each(function (_, codeline) {
+        // unstable idx because lines shift!
+        widget.updateAriaLabel(codeline, false);
+
+        /// definitions for arrow key handling /////////////////////////////
+        const handleVerticalArrowKeys = (e) => {
+          const children = $(codeline.parentElement).children();
+          const index = children.index(codeline);
+          const middleChild = 0 < index && index < children.length - 1;
+          if (e.key.endsWith("Up")) {
+            if (e.ctrlKey || e.metaKey) {
+              children.first().focus();
+            } else if (middleChild || index == children.length - 1) {
+              e.preventDefault();
+              const child = children.eq(index - 1);
+              if (e.altKey) {
+                child.insertAfter(codeline);
+              } else {
+                // focus on first blank if present
+                const blank = child.find("input.parsons-blank").first();
+                (widget.autoEnterBlank && blank.length ? blank : child).focus();
+              }
+            }
+            e.preventDefault();
+            return true;
+          }
+          if (e.key.endsWith("Down")) {
+            if (e.ctrlKey || e.metaKey) {
+              children.last().focus();
+            } else if (middleChild || index == 0) {
+              e.preventDefault();
+              const child = children.eq(index + 1);
+              if (e.altKey) {
+                child.insertBefore(codeline);
+              } else {
+                // focus on first blank if present
+                const blank = child.find("input.parsons-blank").first();
+                (widget.autoEnterBlank && blank.length ? blank : child).focus();
+              }
+            }
+            e.preventDefault();
+            return true;
+          }
+          return false;
+        };
```

Comment:
In a bid to reduce the code duplication between the up/down keys, I spent an embarrassing amount of time refactoring this (and made minimal effect lol). In the process, I think I enabled moving to the edges (top/bottom) of a tray. I'm writing this because I'm not sure if the functionality achieved (bulleted below) is exactly desireable.

Functionality differences:
* only `preventDefault()` if we make a change to code order/selection
    * i.e. if the user presses "up" at the top of a tray, we don't `preventDefault()`
* can move to top/bottom of tray with Ctrl+Alt+Up/Down

```js
const handleVerticalArrowKeys = (e) => {
  const children = $(codeline.parentElement).children();

  const isUpKey = e.key.endsWith("Up");
  const isDownKey = e.key.endsWith("Down");

  const currIndex = children.index(codeline);

  const isFirst = 0 == currIndex;
  const isLast = currIndex == children.length - 1;
  const hasTarget = (isUpKey && !isFirst) || (isDownKey && !isLast);

  let targetIdx;
  if (e.ctrlKey || e.metaKey) {
    targetIdx = isUpKey ? 0 : children.length - 1;
  } else {
    targetIdx = currIndex + (isUpKey ? 1 : -1);
  }

  if ((isUpKey || isDownKey) && hasTarget) {
    e.preventDefault(); // only prevent key action if we do something

    const target = children.eq(targetIdx);
    if (e.altKey) {
      target.insertAfter(codeline);
    } else {
      // focus on first blank if present
      const blank = target.find("input.parsons-blank").first();
      (widget.autoEnterBlank && blank.length ? blank : target).focus();
    }

    return true;
  }
  return false;
};
```